modAionImpl/src/org/aion/zero/impl/sync/msg/RequestBlocks.java

```diff
@@ -0,0 +1,206 @@
+            if (list.get(0) instanceof RLPList) {
+                list = (RLPList) list.get(0);
+            } else {
+                return null;
+            }
+
+            if (list.size() != 4) {
+                return null;
+            } else {
+                // decode the requested type (number or hash)
+                byte value = new BigInteger(1, list.get(0).getRLPData()).byteValue();
+                if (value < 0 || value > 1) {
+                    return null;
+                }
+                boolean isNumber = value == 1;
+
+                // decode the number of blocks requested
+                int count = new BigInteger(1, list.get(2).getRLPData()).intValue();
+                if (count <= 0) {
+                    return null;
+                }
+
+                // decode the requested ordering of the block range
+                value = new BigInteger(1, list.get(3).getRLPData()).byteValue();
+                if (value < 0 || value > 1) {
+                    return null;
+                }
+                boolean descending = value == 1;
+
+                if (isNumber) {
+                    // decode the start block
+                    long start = new BigInteger(1, list.get(1).getRLPData()).longValue();
+                    if (start <= 0) {
+                        return null;
+                    }
+                    // number based request
+                    return new RequestBlocks(start, count, descending);
+                } else {
+                    byte[] start = list.get(1).getRLPData();
+                    if (start.length != HASH_SIZE) {
+                        return null;
+                    }
+                    // hash based request
+                    return new RequestBlocks(start, count, descending);
+                }
+            }
+        }
+    }
+
+    @Override
+    public byte[] encode() {
+        return RLP.encodeList(
+                RLP.encodeByte(isNumber ? (byte) 1 : (byte) 0),
+                isNumber
+                        ? RLP.encode(
+                                startHeight.longValue()) //  encodeLong has non-standard encoding
+                        : RLP.encodeElement(startHash),
+                RLP.encodeInt(count),
+                RLP.encodeByte(isDescending ? (byte) 1 : (byte) 0));
+    }
+
+    public boolean isNumber() {
+        return isNumber;
+    }
+
+    /**
+     * Returns the height of the first block in the requested range.
+     *
+     * @return the height of the first block in the requested range
+     * @implNote The height is set to {@code null} when the hash was used in creating the request.
+     *     It should only be requested when {@link #isNumber()} is {@code true}.
+     */
+    public Long getStartHeight() {
+        return startHeight;
+    }
+
+    /**
+     * Returns the height of the first block in the requested range.
+     *
+     * @return the height of the first block in the requested range
```

Comment:
should say hash instead of height i think

Reply:
done